controller/konnect/reconciler_controlplaneref.go

```diff
@@ -0,0 +1,239 @@
+	case *configurationv1alpha1.KongVault:
+		if e.Spec.ControlPlaneRef == nil {
+			return none
+		}
+		return mo.Some(*e.Spec.ControlPlaneRef)
+	case *configurationv1alpha1.KongKey:
+		if e.Spec.ControlPlaneRef == nil {
+			return none
+		}
+		return mo.Some(*e.Spec.ControlPlaneRef)
+	case *configurationv1alpha1.KongKeySet:
+		if e.Spec.ControlPlaneRef == nil {
+			return none
+		}
+		return mo.Some(*e.Spec.ControlPlaneRef)
+	case *configurationv1alpha1.KongDataPlaneClientCertificate:
+		if e.Spec.ControlPlaneRef == nil {
+			return none
+		}
+		return mo.Some(*e.Spec.ControlPlaneRef)
+	default:
+		return none
+	}
+}
+
+// handleControlPlaneRef handles the ControlPlaneRef for the given entity.
+// It sets the owner reference to the referenced ControlPlane and updates the
+// status of the entity based on the referenced ControlPlane status.
+func handleControlPlaneRef[T constraints.SupportedKonnectEntityType, TEnt constraints.EntityType[T]](
+	ctx context.Context,
+	cl client.Client,
+	ent TEnt,
+) (ctrl.Result, error) {
+	cpRef, ok := getControlPlaneRef(ent).Get()
+	if !ok {
+		return ctrl.Result{}, nil
+	}
+
+	switch cpRef.Type {
+	case configurationv1alpha1.ControlPlaneRefKonnectNamespacedRef:
+		cp := konnectv1alpha1.KonnectGatewayControlPlane{}
+		// TODO(pmalek): handle cross namespace refs
+		nn := types.NamespacedName{
+			Name:      cpRef.KonnectNamespacedRef.Name,
+			Namespace: ent.GetNamespace(),
+		}
+		// Set namespace of control plane when it is non-empty. Only applyies for cluster scoped resources (KongVault).
+		if ent.GetNamespace() == "" && cpRef.KonnectNamespacedRef.Namespace != "" {
+			nn.Namespace = cpRef.KonnectNamespacedRef.Namespace
+		}
+		if err := cl.Get(ctx, nn, &cp); err != nil {
+			if res, errStatus := patch.StatusWithCondition(
+				ctx, cl, ent,
+				konnectv1alpha1.ControlPlaneRefValidConditionType,
+				metav1.ConditionFalse,
+				konnectv1alpha1.ControlPlaneRefReasonInvalid,
+				err.Error(),
+			); errStatus != nil || !res.IsZero() {
+				return res, errStatus
+			}
+			if k8serrors.IsNotFound(err) {
+				return ctrl.Result{}, ReferencedControlPlaneDoesNotExistError{
+					Reference: nn,
+					Err:       err,
+				}
+			}
+			return ctrl.Result{}, err
+		}
+
+		// Do not continue reconciling of the control plane has incompatible cluster type to prevent repeated failure of creation.
+		// Only CLUSTER_TYPE_CONTROL_PLANE and CLUSTER_TYPE_HYBRID are supported.
+		if cp.Spec.ClusterType != nil &&
+			!lo.Contains(compatibleControlPlaneClusterTypes, *cp.Spec.ClusterType) {
+			if res, errStatus := patch.StatusWithCondition(
+				ctx, cl, ent,
+				konnectv1alpha1.ControlPlaneRefValidConditionType,
+				metav1.ConditionFalse,
+				konnectv1alpha1.ControlPlaneRefReasonInvalid,
+				fmt.Sprintf("Attaching to ControlPlane %s with cluster type %s is not supported", nn, *cp.Spec.ClusterType),
+			); errStatus != nil || !res.IsZero() {
+				return res, errStatus
+			}
+			return ctrl.Result{}, nil
+		}
```

Comment:
~I believe that instead of doing it here we should block the reconciliation at the watch functions level. So ideally we'd add these predicates in https://github.com/Kong/gateway-operator/blob/7d67f04feaafc31146b02b87a3fcd363014319cc/controller/konnect/watch.go#L104-L117. That would require the `client.Client` to be provided there.~

⬆️ That won't allow us to add status condition.

👍